_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,193 @@
+
+**Most methods with a narrow responsibility shouldn't have side effects and shouldn't depend on the state of the class**, which enables sharing and calling them from any place. In other words, Single Responsibility Principle nudges us toward [functional programming](https://en.wikipedia.org/wiki/Functional_programming) style.
+
+***
+Pro tip: thinking about responsibilities helps to notice [unrelated subproblems](https://learning.oreilly.com/library/view/the-art-of/9781449318482/ch10.html) hiding in our methods and classes. When we extract them, we can then see opportunities to reuse them in other places. Moving unrelated subproblems out of the way keeps a component at the [single level or abstraction](http://principles-wiki.net/principles:single_level_of_abstraction), which makes easier to understand the logic of the component.
+***
+
+### Testability
+
+**It's easier to test methods and classes with focused, independent concerns.** This is what [Humble Object](http://xunitpatterns.com/Humble%20Object.html) pattern is all about. Let's continue playing with `TransactionProcessor`:
+```
+class TransactionProcessor {
+    
+  void processTransaction(Transaction t) {
+    boolean isFraud;
+    // logic detecting that the transaction is fraud, many
+    // line of code omitted
+    ...
+    
+    if (isFraud) {
+      logger.log("Detected fraud transaction {}", t);
+      alertingService.sendAlert(new FraudTransactionAlert(t));
+    }
+  }
+}
+```
+In this variant, there is no separate `isFraud()` method. `processTransaction()` conflates the fraud detection and the reporting logic. Then, to test the fraud detection, we either need to mock the `alertingService`, which pollutes the test code with boilerplate, or to intercept and check the logging output, which is also cumbersome, and may also hinder our ability to execute tests in parallel.
+
+It's simpler to test a separate `isFraud()` method. But we would still need to construct a `TransactionProcessor` object and to pass some dummy `Logger` and `AlertingService` objects into it. So it's still easier to test the variant with `FraudDetection` class. Notice that to test the intermediate version without separate `FraudDetection` class, we often find ourselves changing the visibility of the method under test (`isFraud()`, in this example) from private to default (package-private). **Use changing visibility of a method and `@VisibleForTesting` annotation as clues to think if we should split the responsibilities of the enclosing class.**
+
+The article about domain-oriented observability referenced above also explains how extracting observability into a separate class (like `TransactionInstrumentation`) [enables clearer, more focused tests](https://martinfowler.com/articles/domain-oriented-observability.html#DomainProbesEnableCleanerMore-focusedTests).
```

Comment:
Which article? Can you link it again?

Reply:
Shortened this sentence without a reference to the article, just link (the article is under the link).